### PR TITLE
[google_maps_flutter] google_maps_flutter_platform_interface migrated to support nnbd

### PR DESCRIPTION
**Pull request by thisisamank**
google_maps_flutter_platform_interface migrated to support nnbd according to issue #75236

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by thisisamank** [comment]
@stuartmorgan

File: packages/android_alarm_manager/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+##  0.5.0-null-safety
+
+* Migrate to null safety.
+
```
Comment (stuartmorgan) [inline packages/android_alarm_manager/CHANGELOG.md:4]: Please revert the android_alarm_manager changes from this PR.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 2]
```diff
@@ -32,7 +32,7 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
 
   /// Accesses the MethodChannel associated to the passed mapId.
   MethodChannel channel(int mapId) {
-    return _channels[mapId];
+    return _channels[mapId]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:35]: Why is this safe?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 3]
```diff
@@ -43,19 +43,19 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
-    MethodChannel channel;
+    MethodChannel? channel;
     if (!_channels.containsKey(mapId)) {
       channel = MethodChannel('plugins.flutter.io/google_maps_$mapId');
       channel.setMethodCallHandler(
           (MethodCall call) => _handleMethodCall(call, mapId));
       _channels[mapId] = channel;
     }
-    return channel.invokeMethod<void>('map#waitForMap');
+    return channel!.invokeMethod<void>('map#waitForMap');
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:53]: This does not appear to be safe; the local variable is never initialized if _channels.containsKey(mapId) is true (which looks like a bug)

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 4]
```diff
@@ -134,7 +134,7 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
       case 'camera#onMove':
         _mapEventStreamController.add(CameraMoveEvent(
           mapId,
-          CameraPosition.fromMap(call.arguments['position']),
+          CameraPosition.fromMap(call.arguments['position'])!,
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:137]: You're using `!` a lot here in ways that don't appear to be safe. There are obvious codepaths that can return null from these methods. Extensive use of `!` is generally an indication that the migration is incomplete or incorrect.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 5]
```diff
@@ -422,13 +422,13 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
   Future<ScreenCoordinate> getScreenCoordinate(
     LatLng latLng, {
-    @required int mapId,
+    required int mapId,
   }) async {
-    final Map<String, int> point = await channel(mapId)
+    final Map<String, int>? point = (await channel(mapId)
         .invokeMapMethod<String, int>(
-            'map#getScreenCoordinate', latLng.toJson());
+            'map#getScreenCoordinate', latLng.toJson()));
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:429]: Why did you add () here?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/bitmap.dart [hunk 6]
```diff
@@ -114,10 +114,10 @@ class BitmapDescriptor {
       _fromAssetImage,
       assetBundleImageKey.name,
       assetBundleImageKey.scale,
-      if (kIsWeb && configuration?.size != null)
+      if (kIsWeb && configuration.size != null)
         [
-          configuration.size.width,
-          configuration.size.height,
+          configuration.size!.width,
+          configuration.size!.height,
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/bitmap.dart:120]: You could make this slightly safer against future changes by making a local variable 
`final size = configuration.size` and then changing all three uses here to `size`. That way the analyzer will know it's not null, and you can eliminate the `!`s.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 7]
```diff
@@ -73,13 +73,13 @@ class CameraPosition {
...
   /// Mainly for internal use.
-  static CameraPosition fromMap(dynamic json) {
+  static CameraPosition? fromMap(dynamic json) {
     if (json == null) {
       return null;
     }
     return CameraPosition(
       bearing: json['bearing'],
-      target: LatLng.fromJson(json['target']),
+      target: LatLng.fromJson(json['target'])!,
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:82]: How do you know the json actually has this key?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 8]
```diff
@@ -157,7 +157,7 @@ class CameraUpdate {
   /// Returns a camera update that modifies the camera zoom level by the
   /// specified amount. The optional [focus] is a screen point whose underlying
   /// geographical location should be invariant, if possible, by the movement.
-  static CameraUpdate zoomBy(double amount, [Offset focus]) {
+  static CameraUpdate zoomBy(double amount, [Offset? focus]) {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:160]: Optional positional arguments are generally discouraged. @cyanglaz Should we use the fact that this will be a breaking change PR anyway to make this a named parameter?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart [hunk 9]
```diff
@@ -31,7 +31,7 @@ class CircleUpdates {
     final Set<CircleId> currentCircleIds = currentCircles.keys.toSet();
 
     Circle idToCurrentCircle(CircleId id) {
-      return currentCircles[id];
+      return currentCircles[id]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart:34]: Why is this safe?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart [hunk 10]
```diff
@@ -45,7 +45,7 @@ class CircleUpdates {
     /// Returns `true` if [current] is not equals to previous one with the
     /// same id.
     bool hasChanged(Circle current) {
-      final Circle previous = previousCircles[current.circleId];
+      final Circle previous = previousCircles[current.circleId]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart:48]: Same question here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart [hunk 11]
```diff
@@ -79,10 +79,10 @@ class CircleUpdates {
       }
     }
 
-    addIfNonNull('circlesToAdd', serializeCircleSet(circlesToAdd));
-    addIfNonNull('circlesToChange', serializeCircleSet(circlesToChange));
+    addIfNonNull('circlesToAdd', serializeCircleSet(circlesToAdd!));
+    addIfNonNull('circlesToChange', serializeCircleSet(circlesToChange!));
     addIfNonNull('circleIdsToRemove',
-        circleIdsToRemove.map<dynamic>((CircleId m) => m.value).toList());
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/circle_updates.dart:85]: All three of these changes look like they would crash if the relevant variable were null. Are there tests covering this code?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/location.dart [hunk 12]
```diff
@@ -102,13 +102,13 @@ class LatLngBounds {
...
-  static LatLngBounds fromList(dynamic json) {
+  static LatLngBounds? fromList(dynamic json) {
     if (json == null) {
       return null;
     }
     return LatLngBounds(
-      southwest: LatLng.fromJson(json[0]),
-      northeast: LatLng.fromJson(json[1]),
+      southwest: LatLng.fromJson(json[0])!,
+      northeast: LatLng.fromJson(json[1])!,
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/location.dart:111]: How do you know the json is well-formed?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart [hunk 13]
```diff
@@ -32,31 +32,31 @@ class InfoWindow {
...
+  final String? snippet;
 
   /// The icon image point that will be the anchor of the info window when
   /// displayed.
   ///
   /// The image point is specified in normalized coordinates: An anchor of
   /// (0.0, 0.0) means the top left corner of the image. An anchor
   /// of (1.0, 1.0) means the bottom right corner of the image.
-  final Offset anchor;
+  final Offset? anchor;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart:48]: Why is this nullable?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker_updates.dart [hunk 14]
```diff
@@ -31,7 +31,7 @@ class MarkerUpdates {
     final Set<MarkerId> currentMarkerIds = currentMarkers.keys.toSet();
 
     Marker idToCurrentMarker(MarkerId id) {
-      return currentMarkers[id];
+      return currentMarkers[id]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker_updates.dart:34]: Same questions about safety here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker_updates.dart [hunk 15]
```diff
@@ -79,10 +79,10 @@ class MarkerUpdates {
       }
     }
 
-    addIfNonNull('markersToAdd', serializeMarkerSet(markersToAdd));
-    addIfNonNull('markersToChange', serializeMarkerSet(markersToChange));
+    addIfNonNull('markersToAdd', serializeMarkerSet(markersToAdd!));
+    addIfNonNull('markersToChange', serializeMarkerSet(markersToChange!));
     addIfNonNull('markerIdsToRemove',
-        markerIdsToRemove.map<dynamic>((MarkerId m) => m.value).toList());
+        markerIdsToRemove!.map<dynamic>((MarkerId m) => m.value).toList());
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker_updates.dart:85]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polygon_updates.dart [hunk 16]
```diff
@@ -31,7 +31,7 @@ class PolygonUpdates {
     final Set<PolygonId> currentPolygonIds = currentPolygons.keys.toSet();
 
     Polygon idToCurrentPolygon(PolygonId id) {
-      return currentPolygons[id];
+      return currentPolygons[id]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polygon_updates.dart:34]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polygon_updates.dart [hunk 17]
```diff
@@ -79,10 +79,10 @@ class PolygonUpdates {
       }
     }
 
-    addIfNonNull('polygonsToAdd', serializePolygonSet(polygonsToAdd));
-    addIfNonNull('polygonsToChange', serializePolygonSet(polygonsToChange));
+    addIfNonNull('polygonsToAdd', serializePolygonSet(polygonsToAdd!));
+    addIfNonNull('polygonsToChange', serializePolygonSet(polygonsToChange!));
     addIfNonNull('polygonIdsToRemove',
-        polygonIdsToRemove.map<dynamic>((PolygonId m) => m.value).toList());
+        polygonIdsToRemove!.map<dynamic>((PolygonId m) => m.value).toList());
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polygon_updates.dart:85]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polyline_updates.dart [hunk 18]
```diff
@@ -32,7 +32,7 @@ class PolylineUpdates {
     final Set<PolylineId> currentPolylineIds = currentPolylines.keys.toSet();
 
     Polyline idToCurrentPolyline(PolylineId id) {
-      return currentPolylines[id];
+      return currentPolylines[id]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polyline_updates.dart:35]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polyline_updates.dart [hunk 19]
```diff
@@ -80,10 +80,10 @@ class PolylineUpdates {
       }
     }
 
-    addIfNonNull('polylinesToAdd', serializePolylineSet(polylinesToAdd));
-    addIfNonNull('polylinesToChange', serializePolylineSet(polylinesToChange));
+    addIfNonNull('polylinesToAdd', serializePolylineSet(polylinesToAdd!));
+    addIfNonNull('polylinesToChange', serializePolylineSet(polylinesToChange!));
     addIfNonNull('polylineIdsToRemove',
-        polylineIdsToRemove.map<dynamic>((PolylineId m) => m.value).toList());
+        polylineIdsToRemove!.map<dynamic>((PolylineId m) => m.value).toList());
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polyline_updates.dart:86]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay.dart [hunk 20]
```diff
@@ -80,14 +80,14 @@ class TileOverlay {
...
-  final TileProvider tileProvider;
+  final TileProvider? tileProvider;
 
   /// The transparency of the tile overlay. The default transparency is 0 (opaque).
   final double transparency;
 
   /// The tile overlay's zIndex, i.e., the order in which it will be drawn where
   /// overlays with larger values are drawn above those with lower values
-  final int zIndex;
+  final int? zIndex;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay.dart:90]: @cyanglaz Is there a reason there's no default value for zIndex? Seems odd to have it allow null rather than have a default (e.g., 0)

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay_updates.dart [hunk 21]
```diff
@@ -31,7 +31,7 @@ class TileOverlayUpdates {
         currentTileOverlays.keys.toSet();
 
     TileOverlay idToCurrentTileOverlay(TileOverlayId id) {
-      return currentTileOverlays[id];
+      return currentTileOverlays[id]!;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay_updates.dart:34]: Same questions here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay_updates.dart [hunk 22]
```diff
@@ -88,12 +88,12 @@ class TileOverlayUpdates {
...
     addIfNonNull(
-        'tileOverlaysToAdd', serializeTileOverlaySet(_tileOverlaysToAdd));
+        'tileOverlaysToAdd', serializeTileOverlaySet(_tileOverlaysToAdd!));
     addIfNonNull(
-        'tileOverlaysToChange', serializeTileOverlaySet(_tileOverlaysToChange));
+        'tileOverlaysToChange', serializeTileOverlaySet(_tileOverlaysToChange!));
     addIfNonNull(
         'tileOverlayIdsToRemove',
-        _tileOverlayIdsToRemove
+        _tileOverlayIdsToRemove!
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/tile_overlay_updates.dart:96]: And here.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ui.dart [hunk 23]
```diff
@@ -39,13 +39,13 @@ class CameraTargetBounds {
...
   /// A null value means the camera target is unbounded.
-  final LatLngBounds bounds;
+  final LatLngBounds? bounds;
 
   /// Unbounded camera target.
   static const CameraTargetBounds unbounded = CameraTargetBounds(null);
 
   /// Converts this object to something serializable in JSON.
-  dynamic toJson() => <dynamic>[bounds?.toJson()];
+  dynamic toJson() => <dynamic>[bounds!.toJson()];
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ui.dart:48]: You can't assert that something that is documented to allow null is not null.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
I'm not familiar with this plugin's code, so we should get a second pass from someone else when it's further along, but in general there is way too much use of `!` here, including some cases that are obviously unsafe. Can you explain the process you are using to do the migration?

File: packages/android_alarm_manager/CHANGELOG.md (hunk 1 above)
Comment (thisisamank) [reply packages/android_alarm_manager/CHANGELOG.md:4]: @stuartmorgan , Actually I am new to open source, I already had PR of `anroid_alarm_manager` that was being reviewed...I started working on this issue by making a new branch and then pushed the code to upstream. I don't know but it got merged with my last PR. Can you please tell me how to revert this?

**Comment by thisisamank** [comment]
> I'm not familiar with this plugin's code, so we should get a second pass from someone else when it's further along, but in general there is way too much use of `!` here, including some cases that are obviously unsafe. Can you explain the process you are using to do the migration?

Also, I am extremely sorry for messing up things here. This is the way I followed: I changed the dart SDK version from pubspec.yaml and then I started changing the file where there was an error. This has introduced a lot of bugs as you rightly said. I have reverted changes locally on my end and will restart everything again. Can you suggest to me what is a better approach?

File: packages/android_alarm_manager/CHANGELOG.md (hunk 1 above)
Comment (stuartmorgan) [reply packages/android_alarm_manager/CHANGELOG.md:4]: You could either revert the commits on that branch, or you can check out those files from master and commit the changes. Resources like https://git-scm.com/book/en/v2/ and StackOverflow would be good places to learn more about how to manage branches in `git`.

**Comment by stuartmorgan** [comment]
> I changed the dart SDK version from pubspec.yaml and then I started changing the file where there was an error. [...] Can you suggest to me what is a better approach?

Null safety migration is not a purely mechanical process; you need to look at the context of the code and figure out what the right way to handle nulls is in each situation. If you're still learning about the basics of null safety, a large and high-profile plugin is not going to be a good starter project, so I think the best thing to do here is close this PR. A small plugin, or perhaps a personal project where you are familiar with the code in general, would be much better places to learn.
